rust_src/src/keymap.rs

```diff
@@ -614,4 +618,67 @@ pub fn describe_vector_lisp(vector: LispObject, mut describer: LispObject) {
     unbind_to(count, Qnil);
 }
 
+#[no_mangle]
+pub extern "C" fn copy_keymap_1(chartable: LispObject, idx: LispObject, elt: LispObject) {
+    unsafe { Fset_char_table_range(chartable, idx, copy_keymap_item(elt)) };
+}
+
+/// Return a copy of the keymap KEYMAP.
+///
+/// Note that this is almost never needed.  If you want a keymap that's like
+/// another yet with a few changes, you should use map inheritance rather
+/// than copying.  I.e. something like:
+///
+/// (let ((map (make-sparse-keymap)))
+/// (set-keymap-parent map <theirmap>)
+/// (define-key map ...)
+/// ...)
+///
+/// After performing `copy-keymap', the copy starts out with the same definitions
+/// of KEYMAP, but changing either the copy or KEYMAP does not affect the other.
+/// Any key definitions that are subkeymaps are recursively copied.
+/// However, a key definition which is a symbol whose definition is a keymap
+/// is not copied.
+#[lisp_fn]
+pub fn copy_keymap(keymap: LispObject) -> LispObject {
+    let mut keymap = get_keymap(keymap, true, false);
+    let mut tail = list!(Qkeymap);
+    let copy = tail;
+
+    keymap = keymap.as_cons_or_error().cdr(); // Skip the `keymap' symbol.
+
+    while let Some(cons) = keymap.as_cons() {
+        let mut elt = cons.car();
+        if elt.eq(Qkeymap) {
+            break;
+        }
+
+        if elt.is_char_table() {
+            elt = unsafe { Fcopy_sequence(elt) };
+            unsafe { map_char_table(Some(copy_keymap_1), Qnil, elt, elt) };
+        } else if let Some(v) = elt.as_vector() {
+            elt = unsafe { Fcopy_sequence(elt) };
+            let mut v2 = elt.as_vector().unwrap();
+            for (i, obj) in v.iter().enumerate() {
+                v2.set(i, unsafe { copy_keymap_item(obj) });
+            }
+        } else if let Some(cons_1) = elt.as_cons() {
+            let front = cons_1.car();
+            if front.eq(Qkeymap) {
+                // This is a sub keymap
+                elt = copy_keymap(elt);
+            } else {
+                elt = LispObject::cons(front, unsafe { copy_keymap_item(cons_1.cdr()) });
+            }
+        }
+
+        setcdr(tail.as_cons().unwrap(), list!(elt));
+        tail = tail.as_cons().unwrap().cdr();
+        keymap = cons.cdr();
+    }
+
+    setcdr(tail.as_cons().unwrap(), keymap);
```

Comment:
Don't we use `as_cons_or_error` instead of `unwrap` ?

Reply:
Usually. But here we know we have cons. We need an equivalent to XCAR and friends.